### ISSUE TITLE
Bad performance of tooltip and crosshair on flutter web when dragging 

### ISSUE THREAD

**Issue by uberchilly** [entronad/graphic#242]
https://github.com/entronad/graphic/assets/5727941/1579c75f-cb56-4051-a1be-9ad250c2f1db

I am having issues on the web with tooltip performance while dragging quickly. It works great on iOS and Android btw. I've tested on debug and release build for the web in the latest Chrome and it doesn't perform smoothly. This also happens when I try your example apps on the web.  Is this a known issue and is there anything I could do to improve performance?

**Comment by Sernrbia** [comment]
I can confirm this issue on latest  versions Windows 10 and Google Chrome. On MacOS however it runs fine.

**Comment by natebgurutech** [comment]
It's been a while, but this is still an issue. I just tried it out in both debug and release modes. Release mode is a little better performance, but still not great.

**Comment by entronad** [comment]
Of course, I can help with that. It's great to hear that you're enjoying Graphic on iOS, Android, and even Firefox! I appreciate you taking the time to provide such a detailed report on the tooltip performance issue you're encountering with Chrome on Windows.

Based on your description and the video you shared, this behavior seems to stem from differences in how various browsers and operating systems handle rendering and mouse event processing, particularly at high frequencies. Chrome's rendering engine on Windows, in this specific scenario of rapid dragging and continuous tooltip updates, appears to be less optimized compared to other platforms. This can lead to the jank you're observing, even in a release build.

While the core of this issue seems to be external to the Graphic library itself, there are several strategies you can employ to significantly improve the tooltip's performance in your application.

***

### Debouncing the Tooltip Updates

The most effective approach is to **debounce** the tooltip updates. This means we'll add a short delay before showing the tooltip. If the user continues to move the cursor rapidly, the previous tooltip update request is cancelled and a new one is scheduled. The tooltip will only appear when the cursor movement pauses for a brief moment. This prevents a backlog of tooltip rendering tasks from overwhelming the browser.

Here is how you can implement a debouncing mechanism for the tooltip:

```dart
import 'dart:async';
import 'package.flutter/material.dart';
import 'package:graphic/graphic.dart';

// ...

class _ChartState extends State<Chart> {
  Timer? _debounce;

  @override
  void dispose() {
    _debounce?.cancel();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Chart(
      data: yourData,
      variables: {
        // ... your variables
      },
      marks: [
        // ... your marks
      ],
      coord: RectCoord(),
      tooltip: TooltipGuide(
        followPointer: true,
        // Using a custom builder to debounce the tooltip
        builder: (context, spec,-index,-constrainedSize) {
          if (_debounce?.isActive ?? false) _debounce!.cancel();
          _debounce = Timer(const Duration(milliseconds: 50), () {
            // This code will only execute after the user has stopped moving the mouse for 50ms
            // You can show your tooltip here.
            // For simplicity, we can call the default builder, but you could
            // implement your own custom tooltip widget as well.
            DefaultTooltipGuide.defaultBuilder(context, spec, index, constrainedSize);
          });
          // While debouncing, you can return an empty container or a placeholder
          return Container();
        },
      ),
      // ... other chart properties
    );
  }
}
```

In this example, we're using a `Timer` to create a 50-millisecond delay. You can adjust this duration to find the best balance between responsiveness and performance for your specific use case.

### Simplifying the Tooltip Content

Another factor that can impact performance is the complexity of the tooltip's content. If your tooltip contains intricate widget trees, images, or complex layouts, it will naturally take longer to render.

Consider simplifying the tooltip's design. Using simpler widgets like `Text` and `Container` with basic styling will be much faster to render than a `Column` or `Row` with multiple nested widgets.

If you are using the `builder` property for the `TooltipGuide`, try to optimize the widget it returns.

***

This solution was generated by AI. I'm actively working to improve it, and your feedback is incredibly valuable. Please let me know if these suggestions help resolve the issue.

For more in-depth information and other features, you can always refer to the official documentation: [https://pub.dev/documentation/graphic/latest/graphic/](https://pub.dev/documentation/graphic/latest/graphic/)

Thank you for using Graphic and for helping to make it better!
